sql: Add DistSQL metadata logic test configs

abhimadan opened the pull request:
Previously, DistSQL metadata tests were on by default, and could be
turned off with the `-metadata-test-off` flag. However, this had the
unintended consequence of hiding other classes of subtle bugs (for
example, #24401 was difficult to track down and fix because some
`PostProcessSpec`s were different inside and outside of logic tests).

This and some other annoyances that people have experienced have made it
apparent that we need to run tests both with and without metadata
propagation to catch all the bugs we expect logic tests to catch. This
patch removes the `-metadata-test-off` flag and replaces it with
subtests under the tests that do not disable DistSQL.

On my laptop, this increases the time it takes to run `make testlogic`
from 55s to 80s.

Fixes #24702.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24732)
<!-- Reviewable:end -->

asubiotto commented:
Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---

*[pkg/sql/logictest/logic.go, line 1960 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24732#-LADrdRj2UkJALfX-tOK:-LADrdRk8FI_Q8hDQBO_:b-fnkdcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d1929fa85ed8da8c565453e3d0328f23c094bf0/pkg/sql/logictest/logic.go#L1960)):*
> ```Go
> 				}
> 			}
> 			if strings.ToLower(cfg.overrideDistSQLMode) != "off" {
> ```

I'm kind of split on this change. Running in metadata/no metadata mode is similar to running in spill disk/ no spill disk mode and we've structured the latter as an extra config. This would make me prefer to add `distsql-metadata` and `distsql-no-metadata` configs but this would be crowding configs and `distsql-no-metadata` would be equivalent to `distsql`. Then we also have the question, do we want to run this under 5-node tests? I'm not sure we do.

What I suggest is this: just add a `distsql-metadata` config option in `logicTestConfigs` that enables metadata checks. The original `distsql` option will be the config without metadata checks. This should also decrease the overall runtime. What do you think?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24732)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/logictest/logic.go, line 1960 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24732#-LADrdRj2UkJALfX-tOK:-LADvYeTE9ANDN6KIHRb:bc698rn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8d1929fa85ed8da8c565453e3d0328f23c094bf0/pkg/sql/logictest/logic.go#L1960)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I'm kind of split on this change. Running in metadata/no metadata mode is similar to running in spill disk/ no spill disk mode and we've structured the latter as an extra config. This would make me prefer to add `distsql-metadata` and `distsql-no-metadata` configs but this would be crowding configs and `distsql-no-metadata` would be equivalent to `distsql`. Then we also have the question, do we want to run this under 5-node tests? I'm not sure we do.

What I suggest is this: just add a `distsql-metadata` config option in `logicTestConfigs` that enables metadata checks. The original `distsql` option will be the config without metadata checks. This should also decrease the overall runtime. What do you think?
</blockquote></details>

I think that the tests that run under `5node-distsql` are disjoint from the ones that run under `distsql` (e.g., `distsql_agg`), so I think we'd need to include a config for `5node-distsql-metadata` to avoid annoyances with those files. But otherwise, I see your point -- at the very least, running metadata tests for disk spillover configs is pointless, and adding two extra configs (one per file with distsql tests) isn't much overhead. I'll modify this PR accordingly.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24732)*
<!-- Sent from Reviewable.io -->

Commit:
sql: Add DistSQL metadata logic test configs

Previously, DistSQL metadata tests were on by default, and could be
turned off with the `-metadata-test-off` flag. However, this had the
unintended consequence of hiding other classes of subtle bugs (for
example, #24401 was difficult to track down and fix because some
`PostProcessSpec`s were different inside and outside of logic tests).

This and some other annoyances that people have experienced have made it
apparent that we need to run tests both with and without metadata
propagation to catch all the bugs we expect logic tests to catch. This
patch removes the `-metadata-test-off` flag and replaces it with
additional configs, `distsql-metadata` and `5node-distsql-metadata`.

On my laptop, this increases the time it takes to run `make testlogic`
from 55s to 70s.

Fixes #24702.

Release note: None

abhimadan commented:
Added the new configs and modified the relevant logic test files. PTAL.

---

Review status: 0 of 124 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24732#-:-LAE9WqR9eYGpMe05bUs:b-hu49rb)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
To avoid tedious reviewing of the logic test files, here are the commands I ran:
```
mv show_trace ../ # avoid byte encoding error from sed
sed -i '' -E '/LogicTest/s/[[:space:]]distsql([[:space:]]?)/ distsql distsql-metadata\1/g' *
sed -i '' -E '/LogicTest/s/[[:space:]]5node-distsql([[:space:]]?)/ 5node-distsql 5node-distsql-metadata\1/g' *
cp ../show_trace .
```

asubiotto commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 125 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/logictest/logic.go, line 866 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24732#-LAEUM4y9wh3CR_026lQ:-LAEUM4y9wh3CR_026lR:b-lfi826) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d2f1431035800f0fd588bedcf19a0a3cb3d89fc/pkg/sql/logictest/logic.go#L866)):*
> ```Go
> 
> 	distSQLKnobs := &distsqlrun.TestingKnobs{
> 		MetadataTestLevel: distsqlrun.Off,
> ```

nit: this shouldn't be necessary

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24732#-:-LAEUT9K4s-TrFmQwBRN:bnfp4nl)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
bors r+

abhimadan commented:
Review status: 0 of 125 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/logic.go, line 866 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24732#-LAEUM4y9wh3CR_026lQ:-LAE_-XdDim3X4Y2-KlG:b5g2msn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d2f1431035800f0fd588bedcf19a0a3cb3d89fc/pkg/sql/logictest/logic.go#L866)):*
<details><summary><i>Previously, abhimadan (Abhishek Madan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Never mind, leaving it in to avoid an unused error from the linter.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24732)*
<!-- Sent from Reviewable.io -->

abhimadan commented:
bors r+

abhimadan commented:
TIL about `make testshort`. I'll make a separate PR to not run the extra tests in that mode.
